core/schema/modtypes.go

```diff
@@ -0,0 +1,903 @@
+		resultList[i], err = t.underlying.ConvertToSDKInput(ctx, item)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return resultList, nil
+}
+
+func (t *ListType) SourceMod() Mod {
+	return t.underlying.SourceMod()
+}
+
+// CoreModObject represents objects from core (Container, Directory, etc.)
+type CoreModObject struct {
+	coreMod  *CoreMod
+	resolver IDableObjectResolver
+}
+
+var _ ModType = (*CoreModObject)(nil)
+
+func (obj *CoreModObject) ConvertFromSDKResult(_ context.Context, value any) (any, error) {
+	id, ok := value.(string)
+	if !ok {
+		return value, nil
+	}
+	return obj.resolver.FromID(id)
+}
+
+func (obj *CoreModObject) ConvertToSDKInput(ctx context.Context, value any) (any, error) {
+	if _, ok := value.(string); ok {
+		return value, nil
+	}
+	return obj.resolver.ToID(value)
+}
+
+func (obj *CoreModObject) SourceMod() Mod {
+	return obj.coreMod
+}
+
+// UserModObject is an object defined by a user module
+type UserModObject struct {
+	api     *APIServer
+	mod     *UserMod
+	typeDef *core.TypeDef
+
+	// should not be read directly, call Fields() and Functions() instead
+	lazyLoadedFields    []*UserModField
+	lazyLoadedFunctions []*UserModFunction
+	loadErr             error
+	loadLock            sync.Mutex
+}
+
+var _ ModType = (*UserModObject)(nil)
+
+func newModObject(ctx context.Context, mod *UserMod, typeDef *core.TypeDef) (*UserModObject, error) {
+	if typeDef.Kind != core.TypeDefKindObject {
+		return nil, fmt.Errorf("expected object type def, got %s", typeDef.Kind)
+	}
+	obj := &UserModObject{
+		api:     mod.api,
+		mod:     mod,
+		typeDef: typeDef,
+	}
+	return obj, nil
+}
+
+func (obj *UserModObject) TypeDef() *core.TypeDef {
+	return obj.typeDef
+}
+
+func (obj *UserModObject) ConvertFromSDKResult(ctx context.Context, value any) (any, error) {
+	if value == nil {
+		return nil, nil
+	}
+
+	switch value := value.(type) {
+	case string:
+		decodedMap, err := resourceid.DecodeModuleID(value, obj.typeDef.AsObject.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode module id: %w", err)
+		}
+		return obj.ConvertFromSDKResult(ctx, decodedMap)
+	case map[string]any:
+		for k, v := range value {
+			normalizedName := gqlFieldName(k)
+			field, ok, err := obj.FieldByName(ctx, normalizedName)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get field %q: %w", k, err)
+			}
+			if !ok {
+				continue
+			}
+
+			delete(value, k)
+			value[normalizedName], err = field.modType.ConvertFromSDKResult(ctx, v)
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert field %q: %w", k, err)
+			}
+		}
+		return value, nil
+	default:
+		return nil, fmt.Errorf("unexpected result value type %T for object %q", value, obj.typeDef.AsObject.Name)
+	}
+}
+
+func (obj *UserModObject) ConvertToSDKInput(ctx context.Context, value any) (any, error) {
+	if value == nil {
+		return nil, nil
+	}
+
+	// TODO: in theory it's more correct to convert to an ID, but SDKs don't currently handle this correctly.
```

Comment:
@jedevc calling this one out so it's not lost in the noise, does the comment here sound correct? Basically, custom objects are still sent to SDKs as inputs in the form of direct json serialization, whereas core types are sent serialized to their ID.

If so, it's *probably* worth fixing I'd think?

Reply:
Ended up not including code to support this (it's significant complication to handle for a case that may or may not ever happen), but updated the comment and if we ever do need to deal with this, it's fairly straightforward

Reply:
Yup that's fine, the comment looks clear enough, thanks! :tada: